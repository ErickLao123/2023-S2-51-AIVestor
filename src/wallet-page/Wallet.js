import React, { useState, useEffect } from "react";
import {
  View,
  Text,
  TextInput,
  TouchableOpacity,
  FlatList,
} from "react-native";
import {
  getFirestore,
  collection,
  doc,
  getDoc,
  setDoc,
  addDoc,
  query,
  where,
  getDocs,
} from "firebase/firestore";
import { getAuth } from "firebase/auth";
import styles from "./Wallet.style";

const Wallet = () => {
  const [topUpAmount, setTopUpAmount] = useState("");
  const [withdrawAmount, setWithdrawAmount] = useState("");
  const [transactions, setTransactions] = useState([]);
  const [balance, setBalance] = useState(0);

  const auth = getAuth();
  const user = auth.currentUser;
  const userId = user.uid;
  const firestore = getFirestore();
  const walletActivityRef = collection(firestore, "wallet-activity");

  const fetchWalletActivity = async () => {
    try {
      const q = query(walletActivityRef, where("userId", "==", userId));
      const querySnapshot = await getDocs(q);
      const walletActivityData = querySnapshot.docs.map((doc) => doc.data());
      setTransactions(walletActivityData);
    } catch (error) {
      console.error("Error fetching wallet activity:", error);
    }
  };


  useEffect(() => {
    const fetchUserBalance = async (userId) => {
      try {
        const userDocRef = doc(firestore, "users", userId);
        const userDocSnapshot = await getDoc(userDocRef);

        if (userDocSnapshot.exists()) {
          const userData = userDocSnapshot.data();
          const userBalance = userData.balance;

          if (userBalance === undefined) {
            await setDoc(userDocRef, { balance: 0 }, { merge: true });
            setBalance(0);
          } else {
            setBalance(userBalance);
          }
        } else {
          await setDoc(userDocRef, { balance: 0 }, { merge: true });
          setBalance(0);
        }
      } catch (error) {
        console.error("Error fetching user balance:", error);
      }
    };

    fetchUserBalance(userId);
    fetchWalletActivity();
  }, [userId]);

  const handleTopUp = async () => {
    try {
      const newBalance = balance + parseFloat(topUpAmount);
      await setDoc(doc(firestore, "users", userId), { balance: newBalance }, { merge: true });
      await addDoc(walletActivityRef, {
        type: "top_up",
        amount: parseFloat(topUpAmount),
        timestamp: new Date(),
        userId: userId,
      });
      setBalance(newBalance);
      setTopUpAmount("");
  
      // Fetch transactions after top-up
      fetchWalletActivity();
    } catch (error) {
      console.error("Error topping up:", error);
    }
  };

  const handleWithdraw = async () => {
    try {
      const newBalance = balance - parseFloat(withdrawAmount);
      if (newBalance < 0) {
        alert("Insufficient balance for withdrawal.");
        return;
      }
      await setDoc(doc(firestore, "users", userId), { balance: newBalance }, { merge: true });
      await addDoc(walletActivityRef, {
        type: "withdrawal",
        amount: parseFloat(withdrawAmount),
        timestamp: new Date(),
        userId: userId,
      });
      setBalance(newBalance);
      setWithdrawAmount("");

      // Fetch transactions after top-up
      fetchWalletActivity();
    } catch (error) {
      console.error("Error withdrawing:", error);
    }
  };

  return (
    <View style={styles.container}>
      <Text>Balance: ${balance.toFixed(2)}</Text>
      <TextInput
        style={styles.input}
        placeholder="Top Up Amount"
        value={topUpAmount}
        onChangeText={(text) => setTopUpAmount(text)}
        keyboardType="numeric"
      />
      <TouchableOpacity style={styles.button} onPress={handleTopUp}>
        <Text style={styles.buttonText}>Top Up</Text>
      </TouchableOpacity>
      <TextInput
        style={styles.input}
        placeholder="Withdraw Amount"
        value={withdrawAmount}
        onChangeText={(text) => setWithdrawAmount(text)}
        keyboardType="numeric"
      />
      <TouchableOpacity style={styles.button} onPress={handleWithdraw}>
        <Text style={styles.buttonText}>Withdraw</Text>
      </TouchableOpacity>
      <Text>Transactions:</Text>
      <FlatList
        data={transactions}
        keyExtractor={(item, index) => index.toString()}
        renderItem={({ item }) => (
          <View style={styles.transactionItem}>
            <Text>Type: {item.type}</Text>
            <Text>Amount: ${item.amount.toFixed(2)}</Text>
          </View>
        )}
      />
    </View>
  );
};

export default Wallet;
