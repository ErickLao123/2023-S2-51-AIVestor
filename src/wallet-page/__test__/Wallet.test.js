import React from "react";
import renderer from "react-test-renderer";
import { TransactionItem } from "../Wallet";

describe("TransactionItem", () => {
  //   test("should render a transaction item with amount correctly", () => {
  //     const item = {
  //       type: "Buy",
  //       amount: 100.0,
  //     };

  //     const tree = renderer.create(<TransactionItem item={item} />).toJSON();
  //     expect(tree).toMatchSnapshot();
  //   });

  //   test("should render a transaction item without amount correctly", () => {
  //     const item = {
  //       type: "Buy",
  //       symbol: "AAPL",
  //       quantity: 10,
  //       price: 150.0,
  //       totalPrice: 1500.0,
  //       timestamp: { toDate: () => new Date(2023, 1, 15) },
  //     };

  //     const tree = renderer.create(<TransactionItem item={item} />).toJSON();
  //     expect(tree).toMatchSnapshot();
  //   });

  test("should render a transaction item without amount correctly", () => {
    const item = {
      type: "Buy",
      symbol: "AAPL",
      quantity: 10,
      price: 150.0,
      totalPrice: 1500.0,
      timestamp: { toDate: () => new Date(2023, 1, 15) },
    };

    const tree = renderer.create(<TransactionItem item={item} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
