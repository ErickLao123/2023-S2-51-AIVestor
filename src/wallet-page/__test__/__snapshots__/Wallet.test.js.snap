// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionItem should render a transaction item without amount correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "borderColor": "#ccc",
      "borderRadius": 10,
      "borderWidth": 1,
      "marginBottom": 10,
      "padding": 10,
    }
  }
>
  <View
    style={{}}
  >
    <Text
      style={
        {
          "fontSize": 17,
          "fontWeight": "bold",
          "marginBottom": 5,
        }
      }
    >
      Buy
    </Text>
    <Text>
      Symbol: 
      AAPL
    </Text>
    <Text>
      Quantity: 
      10
    </Text>
    <Text>
      Price: $
      150
    </Text>
    <Text>
      Total Price: $
      1500
    </Text>
    <Text>
      Timestamp: 
      2/15/2023, 12:00:00 AM
    </Text>
  </View>
</View>
`;
